plugins/android_boot/fu-android-boot-device.c

```diff
@@ -0,0 +1,410 @@
+/*
+ * Copyright (C) 2022 Dylan Van Assche <me@dylanvanassche.be>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include <fcntl.h>
+#include <string.h>
+
+#include "fu-android-boot-device.h"
+
+#define ANDROID_BOOT_PROC_CMDLINE     "/proc/cmdline"
+#define ANDROID_BOOT_MAX_CMDLINE_SIZE 4096
+#define ANDROID_BOOT_BOOT_SLOT_ARG    "androidboot.slot_suffix"
+#define ANDROID_BOOT_SERIAL_ARG	      "androidboot.serialno"
+#define ANDROID_BOOT_UNKNOWN_VERSION  "0.0.0"
+#define ANDROID_BOOT_SECTOR_SIZE      512
+
+struct _FuAndroidBootDevice {
+	FuUdevDevice parent_instance;
+	const gchar *label;
+	const gchar *boot_slot;
+	const gchar *uuid;
+	const gchar *version;
+	guint64 max_size;
+};
+
+G_DEFINE_TYPE(FuAndroidBootDevice, fu_android_boot_device, FU_TYPE_UDEV_DEVICE)
+
+static gchar *
+fu_android_boot_extract_cmdline_arg_value(const gchar *key)
+{
+	gint rc;
+	gint fd;
+	gchar pcmd[ANDROID_BOOT_MAX_CMDLINE_SIZE];
+	g_auto(GStrv) args = NULL;
+	g_auto(GStrv) key_value = NULL;
+
+	fd = open(ANDROID_BOOT_PROC_CMDLINE, O_RDONLY);
+	rc = read(fd, pcmd, ANDROID_BOOT_MAX_CMDLINE_SIZE);
+	close(fd);
+
+	if (rc < 0) {
+		g_warning("Failed to read %s (%d)", ANDROID_BOOT_PROC_CMDLINE, rc);
+		return NULL;
+	}
+
+	args = g_strsplit(pcmd, " ", -1);
+	for (guint i = 0; args[i] != NULL; i++) {
+		key_value = g_strsplit(args[i], "=", -1);
+
+		if (!(key_value[0] && key_value[1]))
+			continue;
+
+		if (g_strcmp0(key_value[0], key) == 0)
+			return g_strdup(key_value[1]);
+	}
+
+	g_warning("Key %s not found in %s", key, ANDROID_BOOT_PROC_CMDLINE);
+	return NULL;
+}
+
+static void
+fu_android_boot_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FuAndroidBootDevice *self = FU_ANDROID_BOOT_DEVICE(device);
+
+	if (self->boot_slot != NULL)
+		fu_string_append(str, idt, "BootSlot", self->boot_slot);
```

Comment:
I don't think you actually need to check for NULL here, I think fu_string_append skips when the value is NULL. Pls check tho, no coffee yet.

Reply:
Just checked, indeed NULL immediately returns, will fix.